%\VignetteIndexEntry{Oligo Vignette}
%\VignetteDepends{}
%\VignetteKeywords{Expression, SNP, Affymetrix, NimbleGen}
%\VignettePackage{oligo}
\documentclass{article}

\newcommand{\Rfunction}[1]{{\texttt{#1}}}
\newcommand{\Rmethod}[1]{{\texttt{#1}}}
\newcommand{\Rcode}[1]{{\texttt{#1}}}
\newcommand{\Robject}[1]{{\texttt{#1}}}
\newcommand{\Rpackage}[1]{{\textsf{#1}}}
\newcommand{\Rclass}[1]{{\textit{#1}}}
\newcommand{\oligo}{\Rpackage{oligo}}

\begin{document}
\title{Oligo Package Vignette}
\date{March, 2007}
\author{Benilton Carvalho}
\maketitle

<<setup, echo=FALSE, results=hide>>=
options(width=60)
options(continue=" ")
options(prompt="R> ")
@ 

\section{Introduction}

The \oligo package is designed to support all microarray designs
provided by Affymetrix and NimbleGen: expression, tiling, SNP and exon
arrays. With the increase in the density of the current technologies,
\oligo uses the resources offered by the \Rpackage{BufferedMatrix}
packages to handle the feature-level information. As of now,
chip-specific packages are built via \Rpackage{makePlatformDesign}, but
transitioning to the \Rpackage{pdInfoBuilder} package.

\section{Analyzing Affymetrix SNP Arrays}

Genotyping and copy number analysis can be performed using \oligo and you will need:

\begin{itemize}
\item \oligo and its dependencies;
\item Chip specific data package, eg. \Rpackage{pd.mapping50k.xba240}:
  package that contains the array specifications and SNP annotation.
\item CEL files.
\item FIXME CRLMM/CN-CRLMM pkg? genotype regions etc... FIXME
\end{itemize}

We will start by loading the \oligo package and importing the CEL files available on the \Rpackage{sampleDataAffy100K}.

<<getRawData>>=
library(oligo)
library(sampleDataAffy100K)
pathCelFiles <- system.file("celFiles", package="sampleDataAffy100K")
rawData <- read.celfiles(list.celfiles(path=pathCelFiles, full.names=TRUE))
@ 

Apply SNPRMA.

<<snprma>>=
preProcessedData <- snprma(rawData)
@ 

Apply CRLMM

<<crlmm>>=
crlmmOut <- crlmm(preProcessedData, correctionFile="exampleCorrection.rda")
@ 

\end{document}
